src/imitation/algorithms/preference_comparisons.py

```diff
@@ -448,6 +586,99 @@ def __call__(
         return list(zip(iterator, iterator))
 
 
+class ActiveSelectionFragmenter(Fragmenter):
+    """Sample fragments of trajectories based on active selection.
+
+    Actively picks the fragment paris with the highest uncertainty (variance)
+    of rewards/probabilties/predictions from ensemble model.
+    """
+
+    def __init__(
+        self,
+        preference_predictor: PreferencePredictor,
+        base_fragmenter: Fragmenter,
+        fragment_sample_factor: float,
+        uncertainty_on: str = "logits",
+        custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
+    ):
+        """Initialize the active selection fragmenter.
+
+        Args:
+            preference_predictor: a model that predicts the preference of frag A.
+            base_fragmenter: fragmenter instance to get
+                fragment pairs from trajectories
+            fragment_sample_factor: the factor of the number of
+                fragment pairs to sample from the base_fragmenter
+            uncertainty_on: the variable to calculate the variance on.
+                Can be logit|probability|label.
+            custom_logger: Where to log to; if None (default), creates a new logger.
+
+        Raises:
+            ValueError: `uncertainty_on` not in logit|probability|label.
+        """
+        super().__init__(custom_logger=custom_logger)
+        self.preference_predictor = preference_predictor
+        self.base_fragmenter = base_fragmenter
+        self.fragment_sample_factor = fragment_sample_factor
+        # self.preference_predictor = PreferencePredictor()
+        if not (uncertainty_on in ["logit", "probability", "label"]):
+            raise ValueError(f"{uncertainty_on} not supported.")
+        self.uncertainty_on = uncertainty_on
+
+    def __call__(
+        self,
+        trajectories: Sequence[TrajectoryWithRew],
+        fragment_length: int,
+        num_pairs: int,
+    ) -> Sequence[TrajectoryWithRewPair]:
+        # sample a large number (self.fragment_sample_factor*num_pairs)
+        # of fragments from all the trajectories
+        fragment_pairs = self.base_fragmenter(
+            trajectories,
+            fragment_length,
+            int(self.fragment_sample_factor * num_pairs),
+        )
+        var_estimates = []
+        for i, fragment in enumerate(fragment_pairs):
+            frag1, frag2 = fragment
+            trans1 = rollout.flatten_trajectories([frag1])
+            trans2 = rollout.flatten_trajectories([frag2])
+            if self.uncertainty_on == "logit":
+                with th.no_grad():
+                    # NOTE: Should we first sum the rewards
+                    # and then take the var (curr approach)
+                    # or first take the var and then sum the rewards?
+                    _, rew1_var = self._reward_moments(trans1)
+                    _, rew2_var = self._reward_moments(trans2)
+                var_estimates.append(rew1_var + rew2_var)
```

Comment:
I think we might actually want the variance of the *difference* in rewards.

For example, suppose we have a 2-member ensemble, with member 1 identical to member 0 but with a constant shift of +100.

The *difference* in their returns is always the same, and would have variance zero. But the variance of their predictions would be quite high.

Reply:
When I hear variance on logits that's what I think of as well since the difference in the reward that gets plugged into the soft max not the return it's self.

Reply:
Yes, that's correct. When I wrote this, I was trying to use the `reward_moments` method in `RewardEnsemble` and hence ended up summing the variance for the two fragments separately. Thanks for pointing it out!